File: Arrays/4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        nums.sort()
        n=len(nums)
        i=0
        while i<n:
            j=i+1
            while j<n-1:
                u=j+1
                v=n-1
                while u<v:
                    if nums[i]+nums[j]+nums[u]+nums[v]==target:
                        quadruple=[0]*4
                        quadruple[0]=nums[i]
                        quadruple[1]=nums[j]
                        quadruple[2]=nums[u]
                        quadruple[3]=nums[v]
                        res.append(quadruple)
                        temp1=nums[u]
                        temp2=nums[v]
                        while u<v and temp1==nums[u]:
                            u+=1
                        while v>u and temp2==nums[v]:
                            v-=1
                    elif nums[i]+nums[j]+nums[u]+nums[v]<target:
                        u+=1
                    else:
                        v-=1
                temp1=nums[j]
                while j<n and temp1==nums[j]:
                    j+=1
            temp1=nums[i]
            while i<n and temp1==nums[i]:
                i+=1
        return res

File: Arrays/4sum.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function template for C++

class Solution
{
public:
    // arr[] : int input array of integers
    // k : the quadruple sum required
    vector<vector<int>> fourSum(vector<int> &arr, int s)
    {
        // Your code goes here
        sort(arr.begin(), arr.end());
        vector<vector<int>> res;
        int n = arr.size();
        int i = 0;
        while (i < n)
        {
            int j = i + 1;
            while (j < n - 1)
            {
                int k = j + 1;
                int l = n - 1;
                int target = s - arr[i] - arr[j];
                while (k < l)
                {
                    int actual = arr[k] + arr[l];
                    if (actual == target)
                    {
                        res.push_back({arr[i], arr[j], arr[k], arr[l]});
                        int temp1 = arr[k];
                        int temp2 = arr[l];
                        while (k < l && arr[k] == temp1)
                        {
                            k++;
                        }
                        while (k < l && arr[l] == temp2)
                        {
                            l--;
                        }
                    }
                    else if (actual < target)
                    {
                        k += 1;
                    }
                    else
                    {
                        l -= 1;
                    }
                }
                int temp1 = arr[j];

                while (j < n and arr[j] == temp1)
                {
                    j++;
                }
            }
            int temp1 = arr[i];
            while (i < n and temp1 == arr[i])
            {
                i += 1;
            }
        }
        return res;
    }
};

// { Driver Code Starts.
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, k, i;
        cin >> n >> k;
        vector<int> a(n);
        for (i = 0; i < n; i++)
        {
            cin >> a[i];
        }
        Solution ob;
        vector<vector<int>> ans = ob.fourSum(a, k);
        for (auto &v : ans)
        {
            for (int &u : v)
            {
                cout << u << " ";
            }
            cout << "$";
        }
        if (ans.empty())
        {
            cout << -1;
        }
        cout << "\n";
    }
    return 0;
} // } Driver Code Ends